### ISSUE TITLE
Returning type defined via generic `ref`'d parameter fails to compile

### ISSUE THREAD

**Issue by tersec** [nim-lang/Nim#21086]
### Description

When calling a function defined via a generic type parameter used as part of a `ref` return type, compilation fails, even when Nim asserts that it, itself, views the return type and the type of the returned object as equivalent.

```nim
type V = object

# Does not build
proc g(T: type V): ref T =
  static: doAssert T is V
  (ref V)()

# Builds
proc h(T: type V): ref V =
  static: doAssert T is V
  (ref V)()

discard g(V)
```

### Nim Version

Tested with all of:

```
Nim Compiler Version 1.6.10 [Linux: amd64]
Compiled at 2022-11-23
Copyright (c) 2006-2021 by Andreas Rumpf

active boot switches: -d:release
```

```
Nim Compiler Version 1.6.11 [Linux: amd64]
Compiled at 2022-12-12
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 7f90bcf5b4cbe5b7da534dd79bafbb4cafb313fa
active boot switches: -d:release
```

```
Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-12-12
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: 438863601047cc974d831c6c0dd5533e4e3f0dc5
active boot switches: -d:release
```

### Current Output

```text
repro.nim(13, 10) template/generic instantiation of `g` from here
repro.nim(5, 3) Error: type mismatch: got 'ref V' for '
discard
ref V()' but expected 'ref typedesc[V]'
```

### Expected Output

```text
Successful compilation
```

**Comment by Araq** [comment]
Why would 

```nim

type V = object

# Does not build
proc g(T: type V): ref T =
  static: doAssert T is V
  (ref V)()

```

work? `T` is a typedesc, not `V`, so you build a ref to a typedesc.

**Comment by tersec** [comment]
Sure, but until 1.6.0, code such as
```nim
type V = object

proc g[T](x: int|int): ref T =
  static: doAssert T is V
  (ref V)()

discard g[V](4)
```
would not compile. It'd get confused at different categories of generic parameters. The most compatible fix I found with 1.2.x was using a style of `g(T: type, ...)`. 

And something like
```nim
type V = object

proc g(T: type, x: int|int): ref V =
  static: doAssert T is V
  (ref T)()

discard g(V, 4)
```
isn't great, but it's compatible back to 1.2.x.

At some point this will cease being a concern, but until then, Nim idioms which rely on 1.6.x or later aren't really available for Nimbus, so workarounds it is.

**Comment by tersec** [comment]
Don't care about 1.2 anymore.
